Feat/buildings v2

ponderingdemocritus opened the pull request:
Implements big change in the way buildings affect gameplay. Most changes were implemented in the `library_buildings.cairo` and tests have been created for these via protostar.

- Less buildings more effects
- Buildings now decay exponentially over time
- Buildings are now stored as time rather than integers
- Buildings now take up spacial space on the Realm until they decay
- Buildings restrict what players can build (WIP)
- Coefficients are computed based on what buildings a Realm has

The decay function works by first computing the number of buildings:

`number_of_buildings = building_time_stamp - now / building_time` 

This results in an integer which we pass into the decay rate

`decay rate = decay_slope * number_of_buildings`

Then we pass the decay rate to calculate the `effective_buildings` (these are what actual impact statistics)

`let (effective_building_time, _) = unsigned_div_rem((10000 - decay_rate) * time_balance, 10000)`

Then we pass this time to calculate the effective amount of buildings on a Realm. We then use this figure for any calculation around the app.

By doing it this way the more buildings you make, the greater the stats you have, but for a shorter period of time. The next addition will be to exponentially increase the cost of buildings, so a player has to make a choice about how they value their time.

rvorias commented on contracts/settling_game/L03_Buildings.cairo:
```diff
@@ -429,57 +299,120 @@ func get_buildings_unpacked{
+    let (Barracks) = castle_time.read(token_id)
+    let (MageTower) = castle_time.read(token_id)
+    let (ArcherTower) = castle_time.read(token_id)
+    let (Castle) = castle_time.read(token_id)
```
What's the reason we can't pack these into one felt?

ponderingdemocritus replied:
timestamp takes 25 bits, so can only fit 6 in a felt. The plan was to move these into 2 felts:
- Combat
- Economy + Housing

Just leaving this to last

rvorias commented on contracts/settling_game/L03_Buildings.cairo:
```diff
@@ -429,57 +299,120 @@ func get_buildings_unpacked{
+end
+
+@view
+func get_buildings_time{
```
Please add docstrings for new functions

rvorias commented on contracts/settling_game/L09_Relics.cairo:
```diff
@@ -0,0 +1,152 @@
+
+    let (current_relic_owner) = get_current_relic_holder(loser_token_id)
+
+    let (isEq) = uint256_eq(current_relic_owner, loser_token_id)
```
isEq -> is_equal

rvorias commented on contracts/settling_game/L09_Relics.cairo:
```diff
@@ -0,0 +1,152 @@
+    # Check if loser has winners token
+    let (winners_relic_owner) = get_current_relic_holder(winner_token_id)
+
+    let (holdsRelic) = uint256_eq(winners_relic_owner, loser_token_id)
```
holds_relic

rvorias commented on contracts/settling_game/library/library_buildings.cairo:
```diff
@@ -0,0 +1,224 @@
+from contracts.settling_game.utils.constants import DAY, BASE_SQM
+from lib.cairo_math_64x61.contracts.Math64x61 import Math64x61_ln
+
+namespace BUILDINGS:
```
BUILDINGS -> Buildings
(open for discussion)

ponderingdemocritus replied:
I'm for of capitalizing these. It makes it easy to read the module files and see what is a lib import

rvorias replied:
It's just that normally you do this for consts, while this is just called as `BUILDINGS.foo()`.

rvorias commented on contracts/settling_game/library/library_buildings.cairo:
```diff
@@ -0,0 +1,224 @@
+    end
+
+    # returns time to add
+    func get_final_time{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
same comment about time, here it is a bit confusing. Maybe something like "integrity" or "life" like you used further on

rvorias commented on contracts/settling_game/library/library_calculator.cairo:
```diff
@@ -43,4 +47,58 @@ namespace CALCULATOR:
+    end
+
+    # Returns coefficient for troop production in bp
+    func get_troop_coefficient{syscall_ptr : felt*, range_check_ptr}(
```
so right now building a mage tower also aids archer troop production?

ponderingdemocritus replied:
correct. There is only 1 coefficient that impacts the overall squad.

rvorias reviewed (commented):
Just some formatting + the naming of "time"

ponderingdemocritus commented on contracts/settling_game/L03_Buildings.cairo:
```diff
@@ -132,48 +136,51 @@ func build{
 
-    # BUILD
-    build_buildings(token_id, building_id)
+    let (buildings_time : RealmBuildings) = get_buildings_time(token_id)
```
agreed - integrity makes sense to me

ponderingdemocritus commented on contracts/settling_game/library/library_buildings.cairo:
```diff
@@ -0,0 +1,224 @@
+    RealmBuildingsSize,
+    BuildingsTime,
+    BuildingsDecaySlope,
+)
```
ok will leave

ponderingdemocritus commented on contracts/settling_game/library/library_buildings.cairo:
```diff
@@ -0,0 +1,224 @@
+
+        let idx = building_id - 1
+
+        let (type_label) = get_label_location(building_decay_slope_bp)
```
just below this, there is a label with a array. It is to select this label as an array

ponderingdemocritus commented on contracts/settling_game/proxy/library.cairo:
```diff
@@ -0,0 +1,110 @@
+# SPDX-License-Identifier: MIT
+# OpenZeppelin Contracts for Cairo v0.x.0 (upgrades/library.cairo)
+
```
I think so now. Will update these to new OZ, need to refactor to use protostar files rather than our OZ hack

ponderingdemocritus commented on tests/settling_game/buildings/test_buildings.cairo:
```diff
@@ -0,0 +1,112 @@
+
+    # now - timestamp = param
+    let (base_building_number) = BUILDINGS.calculate_effective_buildings(1, time_balance)
+
```
yeah will add so tests are final, I abuse print too much....

milancermak commented on contracts/settling_game/library/library_buildings.cairo:
```diff
@@ -0,0 +1,400 @@
+
+        let (buildings : felt*) = alloc()
+
+        if building_id == RealmBuildingsIds.House:
```
The if-else branches in these are the same thing. Is that some relic of a refactor or were they meant to do something different?

milancermak commented on contracts/settling_game/library/library_calculator.cairo:
```diff
@@ -43,4 +47,58 @@ namespace CALCULATOR:
         return (happiness)
     end
+
+    func calculate_food{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
No need to implicit args nor alloc_locals in most of the functions in this file.

milancermak reviewed (changes requested):
Great work so far, perfect that you've added tests as well 👏 

Had a look mostly at the library files, then the baby started crying 👶 so I'll at least post what I have so far :)

One general discussion topic - a lot of the functions in `library_buildings` have a name `get_*`, but they only do some calculations. My personal preference is to use `get_*` for `@storage_var` getters (i.e. a `get_` function indicates a read from storage). No strong opinion though, don't want to bikeshed on this, just curious what you think.

rvorias commented:
re:get

I think this is a good idea, also saw some "fetch_" prefixes in the rest of the codebase (flag those)

ponderingdemocritus commented on tests/settling_game/buildings/test_buildings.cairo:
```diff
@@ -0,0 +1,197 @@
+    alloc_locals
+
+    let test_buildings = RealmBuildings(
+        TEST_TIMESTAMP,
```
Currently reusing RealmBuildings struct to store the timestamp of the buildings. So this test for checking if the timestamp set by pack buildings = the unpacked timestamp

ponderingdemocritus replied:
But I am going to adjust this to different times for each building so it can be used in other tests

ponderingdemocritus commented:
Perhaps - `calculate_` is the best for this and `get_` is for storage fetches ?

ponderingdemocritus commented:
This commit has turned into a giga commit after I went down the hole of updating the OZ contracts. It had to be done either way, and now it is...

Also included in this PR:
- Moved test types into individual folders
- Removed unused node, maker and scripts
- Updated Module namespace file to conform with other libraries

ponderingdemocritus commented:
- Added protostar path to devcontainer so can use protostar dep system
- Removed hardcoded openzepplin dir

milancermak commented on contracts/loot/ItemConstants.cairo:
```diff
@@ -0,0 +1,746 @@
+    const Hands = 6
+    const Neck = 7
+    const Ring = 8
+end
```
Just a random comment - without checking how this is used, it looks like an ideal candidate for bitmap values, so they can be ANDed / ORed together:

```
const Weapon = 0x1
const Chest = 0x10
const Head = 0x100
# etc.
```

milancermak reviewed (approved):
This PR might have a sentient AI hidden inside. Or the plan for a fusion reactor. Or a proof of Riemann hypothesis. TBH, I don't know all the brilliance that's hidden inside. Just ship it.
